pkg/server/sotw/v3/watches.go

```diff
@@ -73,3 +77,36 @@ func (w *watch) close() {
 		w.cancel()
 	}
 }
+
+// updateSubscriptionResources provides a normalized view of resources to be used in Cache
+// It is also implementing the new behavior of wildcard as described in
+// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return
+func updateSubscriptionResources(req *discovery.DiscoveryRequest, sub *stream.Subscription) {
+	subscribedResources := make(map[string]struct{}, len(req.ResourceNames))
+	explicitWildcard := false
+	for _, resource := range req.ResourceNames {
+		if resource == "*" {
+			explicitWildcard = true
+		} else {
+			subscribedResources[resource] = struct{}{}
+		}
+	}
+
+	if sub.IsWildcard() && len(req.ResourceNames) == 0 && len(sub.SubscribedResources()) == 0 {
```

Comment:
Copy from this convo: https://github.com/DataDog/go-control-plane/pull/6#discussion_r1453348775

Does that properly handle the case of unsubscribing from explicit wildcard? Probably this is never used, but still I think this violates the protocol spec and should be implemented properly. For example two subsequent requests:

Req(resources: ["*"]) -> explicit wildcard subscription -> IsWildcard() = true
...
Req(resources: []) -> IsWilcard = true, req.ResourceName = 0, GetSubscribedResources() == 0, 
This will fall into this if statement, even though it should not be considered wildcard and instead should be considered unsubscribing from wildcard.

There seem to be no way to implement the legacy wildcard behavior without storing an additional boolean in what you called SubscriptionState:

For historical reasons, if the client sends a request for a given resource type but has never explicitly subscribed to any resource names

You can't know that without storing whether you've ever seen a resource name on the stream, can you? Without storing this explicitly, there is a high chance that you end up mixing up legacy/implicit and explicit wildcards.

Member
Author
@[valerian-roche](https://github.com/valerian-roche) valerian-roche [7 hours ago](https://github.com/DataDog/go-control-plane/pull/6#discussion_r1454387057)
I agree that we do not support the case when the subscription was explicit wildcard and now requests nothing (not wildcard nor resources). The comment below mentions that we do not support this explicit case currently (in sotw or in delta), but I also don't see much value here. I can add this if preferred though

Reply:
Well, it seems a matter of adhering to the protocol, so I think this should arguably be supported, even if implementations do not use this yet.

Reply:
I'll update to handle this case properly